monitor/internal/pod/delete_controller.go

```diff
@@ -0,0 +1,132 @@
+package podmonitor
+
+import (
+	"context"
+	"time"
+
+	"go.aporeto.io/trireme-lib/common"
+	"go.aporeto.io/trireme-lib/monitor/config"
+	"go.aporeto.io/trireme-lib/policy"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// deleteControllerReconcileFunc is the reconciler function signature for the DeleteController
+type deleteControllerReconcileFunc func(context.Context, client.Client, *config.ProcessorConfig, time.Duration, *map[string]client.ObjectKey)
+
+// DeleteController is responsible for cleaning up after Kubernetes because we
+// are missing our native ID on the last reconcile event where the pod has already
+// been deleted. This is also more reliable because we are filling this controller
+// with events starting from the time when we first see a deletion timestamp on a pod.
+// It pretty much facilitates the work of a finalizer without needing a finalizer and
+// also only kicking in once a pod has *really* been deleted.
+type DeleteController struct {
+	client  client.Client
+	handler *config.ProcessorConfig
+
+	deleteCh           chan DeleteEvent
+	reconcileCh        chan struct{}
+	reconcileFunc      deleteControllerReconcileFunc
+	tickerPeriod       time.Duration
+	itemProcessTimeout time.Duration
+}
+
+// NewDeleteController creates a new DeleteController.
+func NewDeleteController(c client.Client, pc *config.ProcessorConfig) *DeleteController {
+	return &DeleteController{
+		client:             c,
+		handler:            pc,
+		deleteCh:           make(chan DeleteEvent, 1000),
+		reconcileCh:        make(chan struct{}),
+		reconcileFunc:      deleteControllerReconcile,
+		tickerPeriod:       5 * time.Second,
+		itemProcessTimeout: 30 * time.Second,
+	}
+}
+
+// GetDeleteCh returns the delete channel on which to queue delete events
+func (c *DeleteController) GetDeleteCh() chan<- DeleteEvent {
+	return c.deleteCh
+}
+
+// GetReconcileCh returns the channel on which to notify the controller about an immediate reconcile event
+func (c *DeleteController) GetReconcileCh() chan<- struct{} {
+	return c.reconcileCh
+}
+
+// Start implemets the Runnable interface
+func (c *DeleteController) Start(z <-chan struct{}) error {
+	backgroundCtx := context.Background()
+	t := time.NewTicker(c.tickerPeriod)
+	m := make(map[string]client.ObjectKey)
+
+	// the poor man's controller loop
+	for {
+		select {
+		case ev := <-c.deleteCh:
+			m[ev.NativeID] = ev.Key
+		case <-c.reconcileCh:
+			c.reconcileFunc(backgroundCtx, c.client, c.handler, c.itemProcessTimeout, &m)
+		case <-t.C:
```

Comment:
So if we miss a delete event then the timer will kick and check with the API server to see if we have the obj, if not delete it, right?
Maybe in the next controller-runtime we can use the direct API-server client while fetching the objects.